bpt_writer_frames.cpp

```diff
@@ -0,0 +1,289 @@
+#include <string>
+#include <stdexcept>
+#include <sys/types.h>
+#include <unistd.h>
+#include <sys/stat.h>
+
+#include <boost/range.hpp>
+#include <boost/scoped_ptr.hpp>
+#include <boost/foreach.hpp>
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string.hpp>
+#include <boost/bind.hpp>
+
+#define CRYPTOPP_ENABLE_NAMESPACE_WEAK 1
+#include <crypto++/md5.h>
+#include <crypto++/hex.h>
+#include <crypto++/files.h>
+
+#include <libtrace/trace.container.hpp>
+
+#include "bpt_writer_frames.hpp"
+#include "bpt_events.hpp"
+
+namespace bpt {
+
+namespace proto = SerializedTrace;
+typedef proto::TraceContainerWriter container_type;
+
+inline std::string string_of_bytes(const bytes_type& bytes) {
+    return std::string(boost::begin(bytes), boost::end(bytes));
+}
+
+struct std_frame_element : boost::noncopyable {
+    enum usage {R, W};
+    std_frame_element(container_type& c,
+                      const operation_event& op)
+        : cont(c) {
+        ::std_frame* sf = frm.mutable_std_frame();
+        sf->set_address(op.addr());
+        sf->set_thread_id(op.tid());
+        sf->set_rawbytes(string_of_bytes(op.bytes()));
+        sf->mutable_operand_pre_list();
+        sf->mutable_operand_post_list();
+    }
+
+    void add(const read_event& e) { add(R, e); }
+    void add(const write_event& e) { add(W, e); }
+    void add(const load_event& e) { add(R, e); }
+    void add(const store_event& e) { add(W, e); }
+    void add(const read_flags_event& e) {
+        BOOST_FOREACH(const flag& f, e.flags()) {
+            if (f.effect() & RD) {
+                bytes_type::value_type b(f.value(e.bytes()));
+                add(R, f.name(), bytes_type(1, b), f.width());
+            }
+        }
+    }
+    void add(const write_flags_event& e) {
+        BOOST_FOREACH(const flag& f, e.flags()) {
+            if (f.effect() & WR) {
+                bytes_type::value_type b(f.value(e.bytes()));
+                add(W, f.name(), bytes_type(1, b), f.width());
+            }
+        }
+    }
+
+    ~std_frame_element() {
```

Comment:
Don't perform any effects in destructor, put this code in `finish` or `write` or `save` member function.